S3Outputer.cc

```diff
@@ -0,0 +1,230 @@
+#include "OutputerFactory.h"
+#include "UnrolledSerializerWrapper.h"
+#include "FunctorTask.h"
+
+using namespace cce::tf;
+
+void S3Outputer::setupForLane(unsigned int iLaneIndex, std::vector<DataProductRetriever> const& iDPs) {
+  auto& s = serializers_[iLaneIndex];
+  switch(index_.serializestrategy()) {
+    case objstripe::SerializeStrategy::kRoot:
+      s = SerializeStrategy::make<SerializeProxy<SerializerWrapper>>();
+      break;
+    case objstripe::SerializeStrategy::kRootUnrolled:
+      s = SerializeStrategy::make<SerializeProxy<UnrolledSerializerWrapper>>();
+      break;
+    default:
+      throw std::runtime_error("S3Outputer: unrecognized serialization strategy");
+  }
+  s.reserve(iDPs.size());
+  for(auto const& dp: iDPs) {
+    s.emplace_back(dp.name(), dp.classType());
+  }
+  if (currentProductStripes_.size() == 0) {
+    currentProductStripes_.resize(iDPs.size());
+    index_.mutable_products()->Reserve(iDPs.size());
+    for(auto const& ss: s) {
+      auto prod = index_.add_products();
+      prod->set_productname(std::string(ss.name()));
+      prod->set_producttype(ss.className());
+      prod->set_flushsize(0);
+    }
+  }
+  // all lanes see same products? if not we'll need a map
+  assert(currentProductStripes_.size() == iDPs.size());
+}
+
+void S3Outputer::printSummary() const {
+  {
+    tbb::task_group group;
+    {
+      auto start = std::chrono::high_resolution_clock::now();
+      TaskHolder th(group, make_functor_task([](){}));
+      flushProductStripes(th, true);
+      flushEventStripe(th, true);
+      serialTime_ += std::chrono::duration_cast<decltype(serialTime_)>(std::chrono::high_resolution_clock::now() - start);
+    }
+    group.wait();
+  }
+
+  if(verbose_ >= 2) {
+    summarize_serializers(serializers_);
+  }
+  std::chrono::microseconds serializerTime = std::chrono::microseconds::zero();
+  for(const auto& lane : serializers_) {
+    for(const auto& s : lane) {
+      serializerTime += s.accumulatedTime();
+    }
+  }
+  std::cout <<"S3Outputer\n  total serial time at end event: "<<serialTime_.count()<<"us\n"
+    "  total non-serializer parallel time at end event: "<<parallelTime_.load()<<"us\n"
+    "  total serializer parallel time at end event: "<<serializerTime.count()<<"us\n";
+}
+void S3Outputer::output(
+    EventIdentifier const& iEventID,
+    SerializeStrategy const& iSerializers,
+    TaskHolder iCallback
+    ) const
+{
+  using namespace std::string_literals;
+  if(verbose_ >= 2) {
+    std::cout <<"   run:"s+std::to_string(iEventID.run)+" lumi:"s+std::to_string(iEventID.lumi)+" event:"s+std::to_string(iEventID.event)+"\n"<<std::flush;
+  }
+  auto sev = currentEventStripe_.add_events();
+  sev->set_offset(eventGlobalOffset_++);
+  sev->set_run(iEventID.run);
+  sev->set_lumi(iEventID.lumi);
+  sev->set_event(iEventID.event);
+
+  auto p = std::begin(currentProductStripes_);
+  auto pi = index_.mutable_products()->begin();
+  for(const auto& s : iSerializers) {
```

Comment:
Rethinking this, I wonder if I made each product own a SerialTaskQueue, so that `S3Outputer::output` synchronously adds a product to the output queue so all products are in the same event order, but allow the product copy to output buffers (eventually with streaming compression, I think) to happen in parallel?

Reply:
In the case of using a `tbb::concurrent_queue`, should I just start a `group()->run(...)` call after each push even though the task group may not end up popping the product for the same lane that I pushed? I'm not clear exactly how important it is to keep tasks in a group. Would it be better to have a special task group for these product collation steps?

Reply:
Given you are trying to queue tasks in a given order, the SerialTaskQueue probably would be easier to use then some combination of `tbb::concurrent_queue` and `task_group::run`